### PR TITLE
Feature/types

### PR DESCRIPTION
**Pull request by tricinel**
Right, so I think this is it. A few changes in there, including:

- removing `fetch-json.js` and replacing it with the `fetch-json.ts` version
- adapting the gulp build task and the npm scripts
- adding the `files` prop in `package.json` so that the all the files in `dist` are included in the published package
- minor changes in the typescript config

The tests are green 👍

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,15 +47,17 @@
       "nonbsp": false
    },
    "scripts": {
-      "pretest": "jshint *.js spec docs && gulp lint-html && gulp build-dist",
-      "test": "mocha --timeout 10000 spec/*.js"
+      "pretest": "jshint *.js spec docs && gulp lint-html && npm run build",
+      "test": "mocha --timeout 10000 spec/*.js",
+      "build": "tsc && gulp build-dist",
+      "typecheck": "tsc --noEmit"
```
Comment (dpilafian) [inline package.json:53]: When does the `"typecheck"` script get called?

File: tsconfig.json [hunk 2]
```diff
@@ -1,15 +1,11 @@
...
-    "./*.ts"
-  ],
   "compilerOptions": {
     "rootDir": "./",
-    "baseUrl": "./",
-    "target": "es2017", // Node.js 8
-    "module": "commonjs",
+    "outDir": "dist",
+    "alwaysStrict": false,
+    "target": "es5", // Node.js 8
```
Comment (dpilafian) [inline tsconfig.json:6]: Node 8 is not maintained anymore, so feel free to bump up to a more modern target.

The build process currently uses babel to transpile and minify, so anyone who needs legacy support can use the minified version: `/dist/fetch-json.min.js`

See `transpileES6` at:
https://github.com/center-key/fetch-json/blob/master/gulpfile.js#L22

Any thoughts on a good way for **tsconfig.json** `"target"` and babel transpile to work together?  I'm not sure what the current best practice is for coordinating these related processes.

File: package.json (hunk 1 above)
Comment (tricinel) [reply package.json:53]: You can call it manually. The `build` task calls the same `tsc` which checks the typing and errors out, so gulp is not called afterwards. I didn't see the point in having a `prebuild` if they don't do something different.

What do you think?

File: gulpfile.js [hunk 3]
```diff
@@ -32,7 +32,7 @@ const task = {
          .pipe(size({ showFiles: true }));
       },
    buildDistribution: () => {
-      return gulp.src('fetch-json.js')
+      return gulp.src('dist/fetch-json.js')
```
[the remark this answers is not in the document]
Comment (tricinel) [reply gulpfile.js:35]: Yeah, I completely agree! I didn't want to mess with the build system too much (initially I wanted to get rollup in there, but decided against it).

We can do a `build` or `.tmp` and git ignore it. See my comment on the gulp tasks below though :)

File: tsconfig.json (hunk 2 above)
Comment (tricinel) [reply tsconfig.json:6]: 👍 I was thinking there surely is a gulp typescript plugin that allows us to run the `ts` version through the regular gulp streams, so we even bypass the need for an intermediary `build` folder.

I know there's a `babel-typescript` preset so we can let babel do the transpilation directly on the ts file.

What do you think?

Comment (dpilafian) [reply tsconfig.json:6]: Bypassing the intermediary step would be awesome.

Comment (tricinel) [reply tsconfig.json:6]: Hi @dpilafian! Sorry for the late reply, I am reworking these bits as discussed :)

Comment (tricinel) [reply tsconfig.json]: @dpilafian I think I solved it now in the latest commit. So gulp now handles everything and there's no more intermediary build step. Have a look and let me know what you think :)

**Pull request merged**